python/ray/autoscaler/v2/scheduler.py

```diff
@@ -0,0 +1,302 @@
+    # nodes.
+    node_type_available_: Dict[NodeType, int] = field(default_factory=dict)
+
+    def __init__(
+        self,
+        nodes: List[SchedulingNode],
+        node_type_available: Dict[NodeType, int],
+        config: ResourceScheduleConfig,
+    ):
+        self._nodes = nodes
+        self._node_type_available = node_type_available
+        self._config = config
+
+    @classmethod
+    def from_schedule_request(
+        cls, req: ScheduleResourceBundlesRequest
+    ) -> "ScheduleContext":
+        """
+        Create a schedule context from a schedule request.
+
+        It will populate the context with the existing nodes and the available node
+        types from the config.
+        """
+        pass
+
+    def get_nodes(self) -> List[SchedulingNode]:
+        # NOTE: We do the deep copy here since the nodes are usually being updated
+        # during the scheduling process. To prevent accidental modification of the
+        # nodes, we return a copy, and callers should modify the context explicitly
+        # with `ctx.update()`
+        return copy.deepcopy(self._nodes)
+
+    def get_node_config(self, node_type: NodeType) -> Optional[NodeTypeConfig]:
+        c = self._config.node_type_configs.get(node_type, None)
+        if c:
+            return copy.deepcopy(c)
+        return None
+
+    def get_sched_config(self) -> ResourceScheduleConfig:
+        return copy.deepcopy(self._config)
+
+    def update(self, new_nodes: List[SchedulingNode]) -> None:
+        """
+        Update the context with the new nodes.
+        """
+        pass
+
+    def get_node_type_available(self) -> Dict[NodeType, int]:
+        return copy.deepcopy(self._node_type_available)
+
+    def get_cluster_shape(self) -> Dict[NodeType, int]:
+        """
+        Return the current cluster shape.
+
+        The cluster shape is a dict of node types and the number of nodes of that type.
+        """
+        cluster_shape = defaultdict(int)
+        for node in self._nodes:
+            cluster_shape[node.node_type] += 1
+        return cluster_shape
+
+    def get_max_num_nodes(self) -> Optional[int]:
+        """
+        Return the max number of nodes allowed in the cluster.
+        """
+        return (
+            self._config.max_num_nodes
+            if self._config.HasField("max_num_nodes")
+            else None
+        )
+
+
+class SimpleResourceScheduler(IResourceScheduler):
+    """
+    A "simple" resource scheduler that schedules resource requests based on the
+    following rules:
+        1. Enforce the minimal count of nodes for each worker node type.
+        2. Enforce the cluster resource constraints.
+        3. Schedule the gang resource requests.
+        4. Schedule the tasks/actor resource requests
+
+    """
+
+    def schedule_resource_bundles(
+        self, request: ScheduleResourceBundlesRequest
+    ) -> ScheduleResourceBundlesReply:
+        pass
```

Comment:
so it's not implemented yet?

Reply:
Implementation is done in the e2e PR https://github.com/ray-project/ray/pull/39851 but not include in this PR.